---
title: "RNA-seq Part 1"
format: html
editor: visual
author: 
  - name: Vinny Chloros
    email: vinny.chloros@snhu.edu
    affiliations: 
      - name: Southern New Hampshire University
date: 3/11/2025
date-modified: today
date-format: long
code-fold: show
toc: true
theme: flatly
---

For this exercise, I'll be learning RNA-seq, a way to quantify gene expression in large data sets created from genetic information. For this notebook, I'll be trying to analyze gene expression in *Drosophila melanogaster* RNA. The data used here is a sample set from the R Bioinformatics Cookbook.

```{r setup}
#| message: false

library(devtools)
library(rbioinfcookbook)

## Remove (#) from the following if you've never installed the following packages

#install.packages("forcats")

#if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

#BiocManager::install("edgeR")

#if (!require("BiocManager", quietly = TRUE))
    #install.packages("BiocManager")

#BiocManager::install("Biobase")

library(forcats)
library(edgeR)
library(Biobase)
```

```{r load data + make matrix}

genes <- count_dataframe[['gene']]

count_dataframe[['gene']] <- NULL

count_matrix <- as.matrix(count_dataframe)

rownames(count_matrix) <- genes
```

```{r phenotypic larval data}

experiments_of_interest <- c("L1Larvae", "L2Larvae")

columns_of_interest <- which(pheno_data[['stage']] %in% experiments_of_interest)
```

```{r group by variable}

grouping <- pheno_data[["stage"]] [columns_of_interest] |>
  forcats::as_factor()
```

```{r column counts}
counts_of_interest <- count_matrix[,counts = columns_of_interest]
```

```{r assemble into object}
count_dge <- edgeR::DGEList(counts = counts_of_interest, group = grouping)
```

```{r visualize}
design <- model.matrix(~grouping)

eset_dge <- edgeR::estimateDisp(count_dge, design)

fit <- edgeR::glmQLFit(eset_dge, design)

result <- edgeR::glmQLFTest(fit, coef=2)

topTags(result)
```
